etcdserver/api/v3rpc/maintenance.go

```diff
@@ -51,6 +53,41 @@ func (ms *maintenanceServer) Defragment(ctx context.Context, sr *pb.DefragmentRe
 	return &pb.DefragmentResponse{}, nil
 }
 
+type snapshotStream struct {
+	gRPCStream pb.Maintenance_SnapshotServer
+}
+
+func (ms *maintenanceServer) Snapshot(sr *pb.SnapshotRequest, srv pb.Maintenance_SnapshotServer) error {
+	snap := ms.bg.Backend().Snapshot()
+	pr, pw := io.Pipe()
+
+	defer func() {
+		pr.Close()
+		pw.Close()
+		if err := snap.Close(); err != nil {
+			plog.Errorf("error closing snapshot (%v)", err)
+		}
+	}()
+
+	go func() { snap.WriteTo(pw) }()
+
+	br := int64(0)
+	buf := make([]byte, 32*1024)
+	for br < snap.Size() {
+		n, err := pr.Read(buf)
```

Comment:
The logic would be the same since readFull also return n and error. but readFull will let us read more and handle temp error: https://golang.org/pkg/io/#ReadFull.